### PR TITLE
History v4: add initial definition

### PR DESCRIPTION
**Pull request by forticulous**
Initial definitions and test for history v4

### CODE REVIEW

File: definitions/npm/history_v4.x.x/test_history_v4.x.x.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+// @flow
```
Comment (AndrewSouthpaw) [inline definitions/npm/history_v4.x.x/test_history_v4.x.x.js]: Put the test file inside the flow version folder. Tests may vary based on flow version.

File: definitions/npm/history_v4.x.x/test_history_v4.x.x.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import createHashHistory from 'history/createHashHistory';
+
+// browser history
+const browser = createBrowserHistory({
+  basename: "",
+  forceRefresh: false,
+  keyLength: 6,
+})
+
+browser.location.pathname
```
Comment (AndrewSouthpaw) [inline definitions/npm/history_v4.x.x/test_history_v4.x.x.js]: Provide expected typings on all these checks. for better and more explicit tests.

File: definitions/npm/history_v4.x.x/test_history_v4.x.x.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  basename: "",
+  hashType: "slash",
+})
+
+hash.location.pathname
+// $ExpectError
+hash.location.state
+
+// $ExpectError
+hash.foo
```
Comment (AndrewSouthpaw) [inline definitions/npm/history_v4.x.x/test_history_v4.x.x.js]: Organize into blocks:

```
// browser history
{
  ...
}

// memory history
...
```

to organize files a bit better.

File: definitions/npm/history_v4.x.x/flow_v0.25.x-/history_v4.x.x.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    basename?: string,
+    hashType: "slash" | "noslash" | "hashbang",
+    getUserConfirmation?: (
+      message: string,
+      callback: (willContinue: boolean) => void,
+    ) => void,
+  };
+
+  declare export default (opts?: HistoryOpts) => HashHistory;
+}
```
Comment (AndrewSouthpaw) [inline definitions/npm/history_v4.x.x/flow_v0.25.x-/history_v4.x.x.js]: It looks like a lot of these concepts are duplicated. You could probably simplify them outside the modules, to be more DRY:

```js
declare type Push = ...

declare module "history/..." {
  declare type HashHistory = {
    ...
    push: Push
  }
}
```

File: definitions/npm/history_v4.x.x/test_history_v4.x.x.js (hunk 1 above)
Comment (forticulous) [reply definitions/npm/history_v4.x.x/test_history_v4.x.x.js]: I tried moving the test file into the specific flow version folder and I got an error on one of the regression tests. It seems like there can't not be a common test file for the all versions

File: definitions/npm/history_v4.x.x/test_history_v4.x.x.js (hunk 2 above)
Comment (forticulous) [reply definitions/npm/history_v4.x.x/test_history_v4.x.x.js]: done

File: definitions/npm/history_v4.x.x/test_history_v4.x.x.js (hunk 3 above)
Comment (forticulous) [reply definitions/npm/history_v4.x.x/test_history_v4.x.x.js]: done

File: definitions/npm/history_v4.x.x/flow_v0.25.x-/history_v4.x.x.js (hunk 4 above)
Comment (forticulous) [reply definitions/npm/history_v4.x.x/flow_v0.25.x-/history_v4.x.x.js]: i extracted out the Action type and reused it

File: definitions/npm/history_v4.x.x/test_history_v4.x.x.js (hunk 1 above)
Comment (forticulous) [reply definitions/npm/history_v4.x.x/test_history_v4.x.x.js]: Hmm it seems to be failing on the regression test `regression-1385_v1.x.x` just intermittently, not related to any changes I made. I guess I got lucky the first time

Comment (forticulous) [reply definitions/npm/history_v4.x.x/test_history_v4.x.x.js]: Ah it looks like there was a change to the error messages in flow 0.66.0 which broke the regression test

Comment (AndrewSouthpaw) [reply definitions/npm/history_v4.x.x/test_history_v4.x.x.js]: Once the issue with flow 0.66.0 resolves itself, can you try moving the test file again?

Comment (forticulous) [reply definitions/npm/history_v4.x.x/test_history_v4.x.x.js]: yeah i'll do that. I'll save it until master is green again for triggering travis

**Comment by GAntoine** [comment]
@forticulous You'll need to rebase this on the latest version of master.

**Comment by forticulous** [comment]
@GAntoine I rebased onto master, but it looks like the regression test failed again on travis. I suspect this is unrelated to any of my changes?

**Comment by GAntoine** [comment]
Yup, a quick job restart should fix it. I'll merge it in once the spec passes.

**Pull request merged**